### ISSUE TITLE
Implement adjoint differentiation mode

### ISSUE THREAD

**Issue by madagra** [pasqal-io/qadence#97]
Adjoint differentiation enables much better memory usage than backpropagation for differentiating quantum circuit. This is particularly important for applications using higher-order derivatives.

Reference paper for application to quantum circuits: https://arxiv.org/abs/2009.02823

## Outline

- [x] prototype of adjoint differentiation `qadence` using PyTorch with limited set of gates. Use of circuit `dagger()` method to invert quantum operations.
- [ ] test with higher-order derivatives
- [ ] generalize the adjoint method to arbitrary circuits
- [ ] **IMPORTANT**: make some benchmarks to show the advantage (memory and/or runtime) of using the adjoint method over the standard backpropagation with `torch.autograd`
- [ ] unit tests and final integration with `qadence` interface by adding an additional differentiation mode
